pulse2percept/utils.py

```diff
@@ -36,61 +37,66 @@
 
 
-def deprecated(arg):
-    """Decorator used to mark functions as deprecated
+class deprecated(object):
+    """Decorator to mark deprecated functions with warning.
 
-    This is a decorator which can be used to mark functions as deprecated.
-    It will result in a warning being emitted when the function is used.
-
-    Adapted from: http://www.artima.com/weblogs/viewpost.jsp?thread=240845
+    Adapted from <https://github.com/scikit-image/scikit-image/blob/master/skimage/_shared/utils.py>,
+    who adapted it from <http://wiki.python.org/moin/PythonDecoratorLibrary>.
 
     Parameters
     ----------
-    alt_func : str, optional
-        The function to use instead of the deprecated one.
-        Be sure to list a string, not a function object.
-        Default: None.
+    alt_func : str
+        If given, tell user what function to use instead.
+    behavior : {'warn', 'raise'}
+        Behavior during call to deprecated function: 'warn' = warn user that
+        function is deprecated; 'raise' = raise error.
+    removed_version : str
+        The package version in which the deprecated function will be removed.
     """
-    def wrap(func):
-        """The function performing the decoration
-
-        This function is perfoming the actual decoration process. It can only
-        have a single argument, which is the function object.
-
-        Parameters
-        ----------
-        func : function object
-            The function that is deprecated.
-        """
-        e_s = "Call to deprecated function %s." % func.__name__
-        if alt_func:
-            e_s += " Use %s instead." % alt_func
-        logging.getLogger(__name__).warn(e_s)
 
-        def wrapped_func(*args, **kwargs):
+    def __init__(self, alt_func=None, behavior='warn', removed_version=None):
+        self.alt_func = alt_func
+        self.behavior = behavior
+        self.removed_version = removed_version
+
+    def __call__(self, func):
+
+        alt_msg = ''
+        if self.alt_func is not None:
+            alt_msg = ' Use ``%s`` instead.' % self.alt_func
+        rmv_msg = ''
+        if self.removed_version is not None:
+            rmv_msg = (' and will be removed in version %s' %
+                       self.removed_version)
+
+        msg = ('Function ``%s`` is deprecated' % func.__name__ +
+               rmv_msg + '.' + alt_msg)
+
+        @functools.wraps(func)
+        def wrapped(*args, **kwargs):
+            if self.behavior == 'warn':
+                logging.getLogger(__name__).warn(msg)
+            elif self.behavior == 'raise':
+                raise RuntimeError(msg)
             return func(*args, **kwargs)
-        return wrapped_func
 
-    alt_func = None
-    if callable(arg):
-        # Direct decoration
-        return wrap(arg)
-    else:
-        # Alternative function given
-        alt_func = arg
-        return wrap
+        # modify doc string to display deprecation warning
+        doc = '**Deprecated function**.' + alt_msg
```

Comment:
Maybe we can add the `.. deprecated:: 0.2` messages in here as well, instead of manually in each docstring?